### PR TITLE
(Closes #1396) Add tree-update mechanism and new ACCClause nodes for data movement.

### LINKED ISSUE #1396: Managing data movement: add information to data directive nodes and mv clause generation to visitor backend
In #1385 Sergi pointed out that the current approach (of generating the code for the clauses of an ACC DATA directive in a `begin_string` method inside the node) isn't really consistent with the visitor/backend code-generation step. He also suggested that we need to capture the variables that will be moved by a DATA directive as part of the Directive node (at the moment, we only generate this list at code-generation time). This will then allow us to examine neighbouring data Directives and optimise between them - i.e. removing unnecessary data movement.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
On reflection, I've realise that `ACCUpdateDirective._sig2refs()` is actually general functionality that will be needed in at least OpenMP too. Possibly it should live in the same place as `Signature`?

**Comment by arporter** [comment]
Ready for a first review now. Probably one for @sergisiso or @rupertford. This approach (if people are happy with it) will need extending to cover OpenMP too at some point - probably simply by moving functionality into a base class. I've just set the integration tests running. I'd not thought of doing that before so I wouldn't be entirely surprised if the NEMO tests hit problems.
Anyway, hopefully it's good enough for a first look.

File: src/psyclone/psyir/nodes/acc_directives.py [hunk 1]
```diff
@@ -737,6 +689,65 @@ def end_string(self):
...
+        readwrites_list = in_outs.signatures_readwrite
+        readwrites = set(readwrites_list)
+        readers_list = sorted(list(set(readers) - readwrites))
+        writers_list = sorted(list(set(writers) - readwrites))
+
+        # We now need to create PSyIR references for all of the signatures
+        # and add them as children of the appropriate clauses.
+        nodes_dict = OrderedDict()
+        for sig in readers_list:
+            sig.create_deep_copy_refs(var_accesses[sig].all_accesses[0].node,
```
Comment (sergisiso) [inline src/psyclone/psyir/nodes/acc_directives.py:728]: This is a continuation of a previous discussion regarding the "create_deep_copy_refs"

I think that logic belongs in this class (or it parent Directive), but probably also some of the things that are currently inside this tree_udpate should be part of it.

Instead of doing the depanalysis/reference_accesses and iteration here an alternative is to bring those also inside the method, I think it makes a more clear API

`Directive._create_data_movement_deep_copy_refs(): (OrderedDict, OrderedDict, OrderedDict)`
or
`Directive._create_data_movement_deep_copy_refs("in"/"out/"inout"): OrderedDict`

This prevents having to pass and typecheck the OrderDict, and also avoids needing to typecheck the node argument which is currently unchecked and it probably should as it is currently a public method.

Also an intermediate method that calls the `create_data_movement_deep_copy_refs` and puts the returned references on the right clause could be introduced, e.g.: `Directive.create/insert_data_movement_clauses()`
The reason for this is that due to how we implemented clauses, I expect this tree_update method to grow, as thing like reduction clauses, Aidan's tasks, possible atomics would all depend on the directive_body and probably need to be updated by this function, so it would be good to have a sub-method for each to make it clear and overview of what it does.

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 2]
```diff
@@ -156,6 +156,8 @@ def __init__(self, loop_types_to_parallelise=None):
         else:
             self._loop_types_to_parallelise = []
         self._clear_messages()
+        # Cache to hold most recently generated VariablesAccessInfo.
+        self._var_access_info = None
```
Comment (sergisiso) [inline src/psyclone/psyir/tools/dependency_tools.py:160]: If you end up using the self.reference_access in the Directive update you could revert this changes. In general having to access the results from a cache after the execution is not an ideal API, and is easy to inadvertently call something in the middle that changes the results.

File: src/psyclone/psyir/tools/read_write_info.py [hunk 3]
```diff
@@ -100,12 +101,27 @@ def write_list(self):
...
-        :returns the list of all signatures written.
+        :returns: the list of all signatures written.
         :rtype: List[:py:class:`psyclone.core.Signature`]
 
         '''
         return [sig for _, sig in self.write_list]
 
+    # -------------------------------------------------------------------------
+    @property
+    def signatures_readwrite(self):
```
Comment (sergisiso) [inline src/psyclone/psyir/tools/read_write_info.py:112]: This may also not be needed if using the self.reference_access but I don't have any problem with it if its useful for other cases.

My only issue is that it feels the same as the functionality in SingleVariableAccessInfo and I am wondering if we are duplicating things.

File: src/psyclone/transformations.py [hunk 4]
```diff
@@ -2646,6 +2654,44 @@ def validate(self, nodes, options):
...
+                            continue
+                        # For an access such as my_struct(ii)%my_array(ji)
+                        # then if we're inside a loop over it we'll actually
+                        # need a loop to do the deep copy:
+                        #   do ii = 1, N
+                        #   !$acc data copyin(my_struct(ii)%my_array)
+                        #   end do
+                        raise TransformationError(
+                            f"Data region contains a structure access "
+                            f"'{DebugWriter()(sref)}' where component "
```
Comment (sergisiso) [inline src/psyclone/transformations.py:2690]: Maybe sref.debug_string() and remove the DebugWriter import?

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter This PR introduces two important new features, the ability to invalidate/regenerate nodes after a change in its children and an improvement on how we generate data movement deep_copy clauses. (It doesn't do what the PR title says if I am not wrong?)

I like the direction on both features, but I proposed high level changes to both APIs as it is worth getting them right before other code starts depending on them. I am just giving a bird's eye view opinion but if you encounter issues that make the current implementation a better fit I am happy to discuss the comments when you are back.

File: src/psyclone/psyir/nodes/acc_clauses.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+class ACCCopyClause(Clause):
+    '''
+    OpenACC copy clause. Specifies a list of variables that are to be copied
+    to the device at the start of the associated region and back to the host
+    at the end.
+
+    '''
+    _children_valid_format = "Reference"
+    _text_name = "CopyClause"
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/acc_clauses.py:52]: Oh yes, removed for all classes in this file.

File: src/psyclone/transformations.py (hunk 4 above)
Comment (arporter) [reply src/psyclone/transformations.py:2689]: Oh yes, done.

File: src/psyclone/psyir/nodes/acc_directives.py [hunk 6]
```diff
@@ -620,6 +622,27 @@ class ACCDataDirective(ACCRegionDirective):
...
+        don't need to change the tree structure).
+
+        :param other: the object we are copying from.
+        :type other: :py:class:`psyclone.psyir.node.Node`
+
+        '''
+        self._disable_tree_update = True
+        super()._refine_copy(other)
+        self._disable_tree_update = False
+
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/acc_directives.py]: That has cleaned things up a lot, thanks.

File: src/psyclone/psyir/tools/read_write_info.py (hunk 3 above)
Comment (arporter) [reply src/psyclone/psyir/tools/read_write_info.py:112]: Reverted.

File: src/psyclone/psyir/tools/dependency_tools.py (hunk 2 above)
Comment (arporter) [reply src/psyclone/psyir/tools/dependency_tools.py:160]: Reverted.

File: src/psyclone/core/signature.py [hunk 7]
```diff
@@ -106,6 +108,83 @@ def __getitem__(self, indx):
     def __str__(self):
         return "%".join(self._signature)
 
+    # ------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/core/signature.py]: Done.

File: src/psyclone/psyir/nodes/acc_directives.py [hunk 8]
```diff
@@ -737,6 +689,65 @@ def end_string(self):
...
+        from psyclone.psyir.tools import DependencyTools
+        dtools = DependencyTools()
+        in_outs = dtools.get_in_out_parameters(self.children)
+        var_accesses = dtools.variable_access_info
+        readers = in_outs.signatures_read
+        writers = in_outs.signatures_written
+        readwrites_list = in_outs.signatures_readwrite
+        readwrites = set(readwrites_list)
+        readers_list = sorted(list(set(readers) - readwrites))
+        writers_list = sorted(list(set(writers) - readwrites))
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/acc_directives.py]: Yes, that is nicer. I've made that change (thanks again). It has however brought up #446 but that can wait for another PR.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/acc_directives.py]: They are gone now.

File: src/psyclone/psyir/nodes/node.py [hunk 9]
```diff
@@ -95,7 +95,10 @@ class ChildrenList(list):
     a callback function that allows the validation of the inserted children.
     Since this is a subclass of the standard list, all operations (e.g. append,
     insert, extend, comparisons, list arithmetic operations) are conserved and
-    making use of the validation.
+    make use of the validation. They also trigger an update of all ancestor
+    nodes so that action can be taken in order to keep the tree consistent when
+    necessary (e.g. to update the data-movement clauses on an OpenACC data
+    region).
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:102]: Oops, bit of an oversight there. Tests added now.

**Comment by arporter** [comment]
I decided that tackling one of the missed lines would actually be easier if I implemented `Call.reference_accesses` so I've done that in #2262. That will need merging before this one.

File: src/psyclone/psyir/nodes/acc_directives.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/acc_directives.py]: Done.

**Comment by arporter** [comment]
This is now blocked by #2262. Once that is on and this branch brought up-to-date it will be ready for review again.
I started looking at getting full coverage of `acc_directives.py` with `acc_directives_test.py` but that's a lot of work and this is urgent so I propose to leave it for now.

**Comment by arporter** [comment]
(You could still review this as however - #2262 will just fix the one missed line reported by CodeCov.)

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter That's much better now, it solves my previous concerns and there are just minor comments to address now.

Th update_signal functionality probably deserves a section on https://psyclone-dev.readthedocs.io/en/latest/psyir.html, maybe after "The parent-child relationship" as it also makes use of the same list methods.

File: src/psyclone/psyir/nodes/acc_clauses.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        :type child: :py:class:`psyclone.psyir.nodes.Node`
+
+        :return: whether the given child and position are valid for this node.
+        :rtype: bool
+
+        '''
+        return isinstance(child, Reference)
+
+
+__all__ = ["ACCCopyClause", "ACCCopyInClause", "ACCCopyOutClause"]
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/acc_clauses.py:123]: Oh yes, done.

File: src/psyclone/tests/psyir/nodes/acc_directives_test.py [hunk 11]
```diff
@@ -390,3 +439,42 @@ def test_accupdatedirective_equality():
     # Check equality fails when different if_present settings
     directive5 = ACCUpdateDirective(sig, "device", if_present=False)
     assert directive1 != directive5
+
+
+def test_accdatadirective_clauses(fortran_reader, fortran_writer):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/acc_directives_test.py:444]: Yep, done.

File: src/psyclone/psyir/nodes/directive.py [hunk 12]
```diff
@@ -42,12 +42,20 @@
...
+
 from psyclone.configuration import Config
+from psyclone.core import Signature, VariablesAccessInfo
 from psyclone.errors import InternalError
 from psyclone.f2pygen import CommentGen
 from psyclone.psyir.nodes.loop import Loop
-from psyclone.psyir.nodes.statement import Statement
+from psyclone.psyir.nodes.reference import Reference
+from psyclone.psyir.nodes.array_of_structures_reference import (
+    ArrayOfStructuresReference)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/directive.py:54]: Oops.

File: src/psyclone/psyir/nodes/directive.py [hunk 13]
```diff
@@ -68,6 +76,100 @@ def clauses(self):
...
+        :returns: a 3-tuple containing dicts describing the reads, writes and
+            readwrites. Each dict contains References indexed by Signatures.
+        :rtype: Tuple[OrderedDict[
+            :py:class:`psyclone.core.Signature`,
+            :py:class:`psyclone.psyir.nodes.Reference`]]
+
+        '''
+        readwrites = OrderedDict()
+        reads = OrderedDict()
+        writes = OrderedDict()
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/directive.py:94]: Good point. It turned out that the docstring was slightly wrong too.

File: src/psyclone/tests/psyir/nodes/directive_test.py [hunk 14]
```diff
@@ -98,14 +100,87 @@ def test_directive_backward_dependence():
...
+    assert isinstance(readwrites[Signature("a")], nodes.Reference)
+    # Subroutine call - the argument is assumed to be read-write.
+    data_trans.apply(sched[5])
+    reads, writes, readwrites = sched[5].create_data_movement_deep_copy_refs()
+    assert not writes
+    # TODO #446 - currently the reference_accesses() method falls through to
+    # the base implementation and so all arguments to a call are marked as
+    # having READ access.
+    if Signature("d") not in readwrites:
+        pytest.xfail("#446 - Call.reference_accesses() needs implementing")
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/directive_test.py:173]: Oh yes :-)

File: src/psyclone/tests/nemo/transformations/openacc/data_directive_test.py [hunk 15]
```diff
@@ -270,6 +271,60 @@ def test_data_ref_read(parser):
...
+        "end program dtype_read\n")
+    schedule = psyir.walk(Schedule)[0]
+    acc_trans = TransInfo().get_trans_name('ACCDataTrans')
+    acc_trans.apply(schedule.children)
+    gen_code = fortran_writer(psyir)
+    assert ("!$acc data copyin(small_holding,small_holding(2)%data), "
+            "copyout(sto_tmp)" in gen_code)
+
+
+def test_multi_array_derived_type_error(fortran_reader, fortran_writer):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/nemo/transformations/openacc/data_directive_test.py:298]: Oops.

File: src/psyclone/psyir/nodes/node.py [hunk 16]
```diff
@@ -293,13 +303,18 @@ def pop(self, index=-1):
...
 
     def reverse(self):
         ''' Extends list reverse method with children node validation. '''
         for index, item in enumerate(self):
             self._validate_item(len(self) - index - 1, item)
-        super(ChildrenList, self).reverse()
+        super().reverse()
+        # Reversing the order of e.g. Statements may alter the read/write
+        # properties of any References.
+        self._node_reference.update_signal()
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:317]: Added and added tests. (We should really put ChildrenList in its own file but that's another matter.)

File: src/psyclone/psyir/nodes/node.py [hunk 17]
```diff
@@ -1502,6 +1524,43 @@ def debug_string(self):
         from psyclone.psyir.backend.debug_writer import DebugWriter
         return DebugWriter()(self)
 
+    def update_signal(self):
+        '''
+        Called whenever there is a change in the PSyIR tree below this node.
+        It is responsible for ensuring that this method does not get
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:1530]: Oops.

File: src/psyclone/psyir/nodes/node.py [hunk 18]
```diff
@@ -1502,6 +1524,43 @@ def debug_string(self):
...
+        recursively and then calls the _update_node() method of the current
+        node. Finally, it propagates the update signal up to the parent node
+        (if any).
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:1533]: Thanks.

File: src/psyclone/psyir/nodes/node.py [hunk 19]
```diff
@@ -1502,6 +1524,43 @@ def debug_string(self):
...
+
+        '''
+        # Ensure that update_signal does not get called recursively. If this
+        # node is still under construction then it might not have the
+        # _disable_tree_update attribute. If so, we also skip doing the update.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:1538]: Oh yes, that works and is much nicer - thanks.

File: src/psyclone/tests/psyir/nodes/node_test.py [hunk 20]
```diff
@@ -1080,6 +1080,63 @@ def test_children_setter():
...
+        Sub-class of Schedule that re-implements update_signal() so that it
+        (configurably) raises a GenerationError when called.
+
+        '''
+        def __init__(self, test_enable=True):
+            # Controls whether or not an exception is raised by update_signal.
+            self._test_enable = test_enable
+            super().__init__()
+
+        def update_signal(self):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/node_test.py:1100]: That proved slightly tricky as raising an exception from _update_node meant that the _disable_tree_update flag was never reset in the calling update_signal routine. As this is a test class I've simply fixed it by resetting that flag myself.

**Comment by arporter** [comment]
Assuming CI happiness,this is ready for another look.

File: src/psyclone/psyir/nodes/node.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:317]: Thanks for adding the missing methods.
I actually like the ChildrenList class being in the same file. It makes clear that it is an internal base Node utility class and not to be used from anywhere else.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter This is almost ready, but I still think that a bit of documentation about the tree update mechanism should be added:

> Th update_signal functionality probably deserves a section on https://psyclone-dev.readthedocs.io/en/latest/psyir.html, maybe after "The parent-child relationship" as it also makes use of the same list methods.

And I guess brining it to master will solve the current test failure?

File: src/psyclone/tests/psyir/nodes/node_test.py [hunk 21]
```diff
@@ -1080,6 +1080,93 @@ def test_children_setter():
     assert statement2.parent is None
 
 
+def test_children_clear():
+    '''Test that the clear() method works correctly for a ChildrenList.'''
+    testnode = Schedule()
+    testnode.addchild(Statement())
+    testnode.addchild(Statement())
+    assert len(testnode.children) == 2
+    testnode.children.clear()
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/node_test.py:1089]: Done.

File: src/psyclone/psyir/nodes/__init__.py [hunk 22]
```diff
@@ -98,6 +100,9 @@
 # this package e.g. 'from psyclone.psyir.nodes import Literal'
 __all__ = [
         'colored',
+        'ACCCopyClause',
+        'ACCCopyInClause',
+        'ACCCopyOutClause',
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/__init__.py:105]: Yes, that's nicer - changed.

**Review by sergisiso — APPROVED** [review]
@arporter Thanks for adding the documentation. All comments have been addressed and all tests and formatting checks are green. (I found a TODO to this issue regarding C++ data range specifications, but I replaced it with a more specific issue #2304). This is ready to merge.
